### ISSUE TITLE
default_branch not master but main

### ISSUE THREAD

**Issue by FernandoMiguel** [gitlabhq/terraform-provider-gitlab#349]
### Terraform Version
Terraform v0.12.26
+ provider.gitlab v2.10.0

### Affected Resource(s)
resource "gitlab_project"

### Terraform Configuration Files
```hcl
resource "gitlab_project" "project" {
  name                                             = var.PROJECT_NAME
  description                                      = var.PROJECT_DESCRIPTION
  path                                             = var.PROJECT_PATH
  approvals_before_merge                           = var.PROJECT_APPROVALS_BEFORE_MERGE
  archived                                         = var.PROJECT_ARCHIVED
  container_registry_enabled                       = var.PROJECT_CONTAINER_REGISTRY_ENABLED
  default_branch                                   = var.PROJECT_DEFAULT_BRANCH
  initialize_with_readme                           = var.PROJECT_INITIALIZE_WITH_README
  issues_enabled                                   = var.PROJECT_ISSUES_ENABLED
  lfs_enabled                                      = var.PROJECT_LFS_ENABLED
  merge_method                                     = var.PROJECT_MERGE_METHOD
  merge_requests_enabled                           = var.PROJECT_MERGE_REQUESTS_ENABLED
  namespace_id                                     = local.GITLAB_GROUP_NAMESPACE_ID
  only_allow_merge_if_all_discussions_are_resolved = var.PROJECT_ONLY_ALLOW_MERGE_IF_ALL_DISCUSSIONS_ARE_RESOLVED
  only_allow_merge_if_pipeline_succeeds            = var.PROJECT_ONLY_ALLOW_MERGE_IF_PIPELINE_SUCCEEDS
  pipelines_enabled                                = var.PROJECT_PIPELINES_ENABLED
  request_access_enabled                           = var.PROJECT_REQUEST_ACCESS_ENABLED
  shared_runners_enabled                           = var.PROJECT_SHARED_RUNNERS_ENABLED
  snippets_enabled                                 = var.PROJECT_SNIPPETS_ENABLED
  tags                                             = var.PROJECT_TAGS
  visibility_level                                 = var.PROJECT_VISIBILITY_LEVEL
  wiki_enabled                                     = var.PROJECT_WIKI_ENABLED
  remove_source_branch_after_merge                 = var.PROJECT_REMOVE_SOURCE_BRANCH_AFTER_MERGE
  # shared_with_groups                               = var.PROJECT_SHARED_WITH_GROUPS
}

variable "PROJECT_DEFAULT_BRANCH" {
  description = "(Optional) The default branch for the project."
  type        = string
  default     = "main"
}
```

### Debug Output
```hcl
Terraform will perform the following actions:

  # gitlab_project.project will be created
  + resource "gitlab_project" "project" {
      + approvals_before_merge                           = 1
      + archived                                         = false
      + container_registry_enabled                       = true
      + http_url_to_repo                                 = (known after apply)
      + id                                               = (known after apply)
      + initialize_with_readme                           = true
      + issues_enabled                                   = false
      + lfs_enabled                                      = true
      + merge_method                                     = "rebase_merge"
      + merge_requests_enabled                           = true
      + name                                             = "foobar"
      + namespace_id                                     = REDACTED
      + only_allow_merge_if_all_discussions_are_resolved = true
      + only_allow_merge_if_pipeline_succeeds            = true
      + pipelines_enabled                                = true
      + remove_source_branch_after_merge                 = true
      + request_access_enabled                           = false
      + runners_token                                    = (known after apply)
      + shared_runners_enabled                           = false
      + snippets_enabled                                 = true
      + ssh_url_to_repo                                  = (known after apply)
      + visibility_level                                 = "private"
      + web_url                                          = (known after apply)
      + wiki_enabled                                     = false
    }

Plan: 1 to add, 0 to change, 0 to destroy.
```

and on a second run
```hcl
  ~ update in-place

Terraform will perform the following actions:

  # gitlab_project.project will be updated in-place
  ~ resource "gitlab_project" "project" {
        approvals_before_merge                           = 1
        archived                                         = false
        container_registry_enabled                       = true
      ~ default_branch                                   = "master" -> "main"
        http_url_to_repo                                 = "https://gitlab.com/REDACTED/foobar.git"
        id                                               = REDACTED
        initialize_with_readme                           = true
        issues_enabled                                   = false
        lfs_enabled                                      = true
        merge_method                                     = "rebase_merge"
        merge_requests_enabled                           = true
        name                                             = "foobar"
        namespace_id                                     = REDACTED
        only_allow_merge_if_all_discussions_are_resolved = true
        only_allow_merge_if_pipeline_succeeds            = true
        path                                             = "foobar"
        pipelines_enabled                                = true
        request_access_enabled                           = false
        runners_token                                    = REDACTED
        shared_runners_enabled                           = false
        snippets_enabled                                 = true
        ssh_url_to_repo                                  = "git@gitlab.com:REDACTED/foobar.git"
        tags                                             = []
        visibility_level                                 = "private"
        web_url                                          = "https://gitlab.com/REDACTED/foobar"
        wiki_enabled                                     = false
    }

Plan: 0 to add, 1 to change, 0 to destroy.
```

### Expected Behavior
for the repo to be created with a `main` default branch

### Actual Behavior
default branch was master and on second apply of tf, it tried to recreate it and failed
```hcl
Error: PUT https://gitlab.com/api/v4/projects/REDACTED: 400 {message: {base: [Could not change HEAD: branch 'main' does not exist]}}

  on main.tf line 18, in resource "gitlab_project" "project":
  18: resource "gitlab_project" "project" {
```

### Steps to Reproduce
Please list the steps required to reproduce the issue, for example:
1. `terraform apply`

**Comment by abitrolly** [comment]
You're using variable for setting the branch.
```
  default_branch                                   = var.PROJECT_DEFAULT_BRANCH
```
And you set `var.PROJECT_DEFAULT_BRANCH` to `main` in the block below. I see no bug here.

**Comment by abitrolly** [comment]
@FernandoMiguel because when project is created the repository is empty and setting `default_branch` on an empty repository will result in an error. GitLab checks for the branch existence prior to setting it.

**Comment by abitrolly** [comment]
@FernandoMiguel ok, I see what you mean. With `initialize_with_readme` the remote repo needs is created with "master" branch, but the `default_branch` setting is applied only on the next invocation.

It seems to me that logically the repo configuration should be done in another resource.

**Comment by FernandoMiguel** [comment]
so what do you purpose @abitrolly ? 
it's not like i can create two resources `gitlab_project` and have the default branch and read me run only there.
I could not init the repo, but that would leave it without a default branch anyway

**Comment by abitrolly** [comment]
@FernandoMiguel just run it twice. I may look into the proper fix, but not before https://github.com/terraform-providers/terraform-provider-gitlab/pull/343 is merged.

**Comment by abitrolly** [comment]
@FernandoMiguel specifically for your repo, you need to make sure the branch `main` exists before setting it. GitLab provider doesn't commit to repos and uses only GitLab API for configuring the project resource, so you need to make commit that creates the branch somehow.

**Comment by FernandoMiguel** [comment]
@abitrolly how do you purpose i run terraform twice on CI?
it's state based... once applied you expect the infra represents what your code was codified with

**Comment by abitrolly** [comment]
You run TF on CI in steps. First provisioning, then configuration, then deploy. Run it in configuration step. Ansible or bash. Whatever your CI allows.

**Comment by abitrolly** [comment]
That's my advice anyway. Maybe somebody else have more time to research a better way. I didn't have this problem.

**Comment by FernandoMiguel** [comment]
i'm going with 
```hcl
variable "PROJECT_INITIALIZE_WITH_README" {
  description = "(Optional) Create master branch with first commit containing a README.md file."
  type        = bool
  default     = false 
  ## making this true will create a branch called master, not the one in default_branch
}
```

**Comment by abitrolly** [comment]
Yup. This way there won't be content pushed to repository as a result of TF calls. And after the repo is non-empty, the Terraform will try to set the branch to `main`.

It might be convenient to be able to create empty repos with GitLab with other default branch, but that's better to be researched in another issue. Because for that GitLab provider needs an ability to make at least one commit to the repo. Maybe the issue can be raised on GitLab side so that API allowed that. Some kind of `init` call.

**Comment by armsnyder** [comment]
I agree with Fernando that when an apply finishes, the upstream state should be fully consistent with the TF resource. This can be accomplished with multiple GitLab API calls in the resource create function if that's how it has to be done.

**Comment by emmasax4** [comment]
@abitrolly The solution I used is `null_resource`s in Terraform. Essentially, create the project with the README with the `master` branch (don't even give a master branch to the project module), and then use `null_resource`s to make a new default branch, update the default branch, unprotect `master`, and delete `master`. You can run it all in one `terraform apply`, and it works every time.

It's smart enough to know not to delete the `master` branch if you set that as your `default_branch` variable, and it also won't unprotect it if you set that as your default branch. It gets a little tricky when it comes to changing default branches repeatedly, so proceed with caution there (almost easier to just do that in the GitLab browser).

```hcl
variable "default_branch" {
  type        = string
  default     = "main"
  description = "The default branch"
}

variable "source_branch" {
  type        = string
  default     = "master"
  description = "The name of the source branch to use when creating new branches"
}

locals {
  branches      = setsubtract(distinct(concat([var.source_branch], [var.default_branch])), ["master"])
  delete_branch = var.default_branch == "master" ? [] : ["master"]
}

resource "null_resource" "gitlab_branch" {
  for_each = toset(local.branches)

  depends_on = [
    gitlab_project.project
  ]

  provisioner "local-exec" {
    command = "curl -H 'PRIVATE-TOKEN: ${local.gitlab_token}' -X POST 'https://gitlab.com/api/v4/projects/${gitlab_project.project.id}/repository/branches?branch=${each.value}&ref=${var.source_branch}'"
  }
}

resource "null_resource" "update_default_branch" {
  depends_on = [
    gitlab_project.project,
    null_resource.gitlab_branch
  ]

  triggers = {
    default_branch = var.default_branch
  }

  provisioner "local-exec" {
    command = "curl -H 'PRIVATE-TOKEN: ${local.gitlab_token}' -X PUT --data 'default_branch=${var.default_branch}' 'https://gitlab.com/api/v4/projects/${gitlab_project.project.id}'"
  }
}

resource "null_resource" "unprotect_master_branch" {
  for_each = setsubtract(toset(local.delete_branch), local.branches)

  depends_on = [
    gitlab_project.project,
    null_resource.gitlab_branch,
    null_resource.update_default_branch
  ]

  triggers = {
    branch_to_unprotect = "master"
  }

  provisioner "local-exec" {
    command = "curl -H 'PRIVATE-TOKEN: ${local.gitlab_token}' -X DELETE 'https://gitlab.com/api/v4/projects/${gitlab_project.project.id}/protected_branches/master'"
  }
}

resource "null_resource" "delete_master_branch" {
  for_each = setsubtract(toset(local.delete_branch), local.branches)

  depends_on = [
    gitlab_project.project,
    null_resource.gitlab_branch,
    null_resource.update_default_branch,
    null_resource.unprotect_master_branch
  ]

  triggers = {
    branch_to_delete = "master"
  }

  provisioner "local-exec" {
    command = "curl -H 'PRIVATE-TOKEN: ${local.gitlab_token}' -X DELETE 'https://gitlab.com/api/v4/projects/${gitlab_project.project.id}/repository/branches/master'"
  }
}
```

**Comment by abitrolly** [comment]
#343 is merged. Now for this issue. There are three levels.

1. Git. It doesn't seem to allow branches (and default branches) on empty repository. This needs to be validated and filled upstream as an RFE.
2. GitLab Git backend. Maybe it can still store extra info with "default branch" even if a repo is empty. This still may create conflict if Git on developer's machines creates another default branch.
3. GitLab API. The solution could be to force initialization with README when "default branch" is specified and a repo is empty. But this still looks like a workaround against Git behaviour.

**Comment by emmasax4** [comment]
Oh yes of course. I don't think it's ideal to have to do a bunch of `null_resource`s all over to get around this. I'd hope that you could initialize a README on whatever branch is passed in as a default branch. But to be honest, I've run into very similar issues with the GitHub Terraform as well, so GitLab isn't alone in their problems (branches not being defined until there's a file).

**Comment by armsnyder** [comment]
I see three options here:
1. Wait for a fix in the upstream API ([this issue](https://gitlab.com/gitlab-org/gitlab/-/issues/26261))
2. Implement @emmahsax's workaround logic into this provider, so that you can set `default_branch` in conjunction with `initialize_with_readme = true` (but still ignore `default_branch` when `initialize_with_readme = false` because you cannot create branches on empty repositories).
3. Same as 2, but if `initialize_with_readme = false` then push an empty initial commit to the project so that the default branch can be set.

Personally I like option 2, since it looks like people only run into this when using `initialize_with_readme = true`, and option 3 has messy side effects.

**Comment by PedDavid** [comment]
Curious on what kind of messy side effects do you experience with 3.?

I wouldn't mind 3. at all as I start all my projects with an empty commit `Create repository` but that's a very personal choice so I'm completely on board with option 2. (seems like the safest choice).

**Comment by abitrolly** [comment]
This can be fixed by upgrading GitLab `git` to 2.28 https://github.blog/2020-07-27-highlights-from-git-2-28/#introducing-init-defaultbranch and make sure the `init.defaultBranch main` setting is passed.

**Comment by abitrolly** [comment]
The upstream API fix is landed - https://gitlab.com/gitlab-org/gitlab/-/issues/26261#note_601841507 - so this issue might be fixed already.

There is also another issue that will allow to set default branch without `initialize_with_readme` https://gitlab.com/gitlab-org/gitlab/-/issues/333426
